File: typescript/src/ui/overlays/help.ts
import P5 from 'p5';
import Colors from '../colors';
import Fonts from '../fonts';
import Keycap from './keycap';
import Overlay from './overlay';
import Spacebar from './spacebar';

export default class Help extends Overlay {
    private keycapLeft: Keycap;
    private keycapRight: Keycap;
    private keycapUp: Keycap;
    private spacebar: Spacebar;

    constructor(p5: P5) {
        super(p5);

        this.keycapLeft = new Keycap(p5, new P5.Vector(650, 285), -90);
        this.keycapRight = new Keycap(p5, new P5.Vector(770, 285), 90);
        this.keycapUp = new Keycap(p5, new P5.Vector(710, 215), 0);
        this.spacebar = new Spacebar(p5, new P5.Vector(710, 365));
    }

    draw() {
        this.p5.push();

        this.p5.fill(Colors.BACKGROUND);
        this.p5.stroke(Colors.EDGE);
        this.p5.strokeWeight(2);
        this.p5.rect(150, 150, this.p5.width - 300, this.p5.height - 300, 10);

        this.p5.fill(Colors.EDGE);
        this.p5.noStroke();
        this.p5.textFont(Fonts.getInstance().fontLight);
        this.p5.textAlign(this.p5.CENTER);
        this.p5.textSize(35);
        this.p5.text('HELP', this.centerX, 210);

        this.p5.textAlign(this.p5.LEFT);
        this.p5.textSize(15);
        this.p5.text('START', 220, 275);
        this.p5.text('PAUSE', 220, 325);
        this.p5.text('HUB', 220, 375);
        this.p5.text('SHOW EGDES', 220, 425);
        this.p5.text('STARFIELD', 220, 475);

        this.p5.textAlign(this.p5.RIGHT);
        this.p5.text('S', 475, 275);
        this.p5.text('P', 475, 325);
        this.p5.text('U', 475, 375);
        this.p5.text('E', 475, 425);
        this.p5.text('X', 475, 475);

        this.keycapLeft.draw();
        this.keycapRight.draw();
        this.keycapUp.draw();

        this.p5.textSize(15);
        this.p5.textAlign(this.p5.CENTER);
        this.p5.text('BOOST', 710, 250);
        this.p5.text('TURN LEFT', 650, 320);
        this.p5.text('TURN RIGHT', 770, 320);
        this.p5.text('SHOOT', 710, 400);

        this.p5.textAlign(this.p5.LEFT);
        this.p5.text('ASTEROID', 600, 450);
        this.p5.text('UFO', 600, 475);
        this.p5.text('NEW LIFE', 600, 500);

        this.p5.text('10 POINTS', 745, 450);
        this.p5.text('50 POINTS', 745, 475);
        this.p5.text('3000 POINTS', 745, 500);

        this.spacebar.draw();

        this.p5.pop();
    }
}

File: rust/src/overlays/help.rs
use web_sys::CanvasRenderingContext2d;

use crate::{colors::Colors, sprites::sprite::CanvasDimension, vector::Vector};

use super::{keycap::Keycap, spacebar::Spacebar};

pub struct Help {
    keycap_left: Keycap,
    keycap_right: Keycap,
    keycap_up: Keycap,
    spacebar: Spacebar,
    center_x: f64,
    canvas: CanvasDimension,
}

impl Help {
    pub fn new(canvas: CanvasDimension) -> Help {
        Help {
            keycap_left: Keycap::new(Vector::new(650.0, 280.0), -90.0),
            keycap_right: Keycap::new(Vector::new(770.0, 285.0), 90.0),
            keycap_up: Keycap::new(Vector::new(710.0, 215.0), 0.0),
            spacebar: Spacebar::new(Vector::new(710.0, 365.0)),
            center_x: canvas.width / 2.0,
            canvas,
        }
    }
}

impl Help {
    pub fn draw(&self, canvas: CanvasRenderingContext2d) {
        canvas.save();

        canvas.set_fill_style(&Colors::Background.value().into());

        canvas.fill_rect(
            150.0,
            150.0,
            self.canvas.width - 300.0,
            self.canvas.height - 300.0,
        );

        canvas.stroke_rect(
            150.0,
            150.0,
            self.canvas.width - 300.0,
            self.canvas.height - 300.0,
        );

        self.keycap_left.draw(canvas.clone());
        self.keycap_right.draw(canvas.clone());
        self.keycap_up.draw(canvas.clone());
        self.spacebar.draw(canvas.clone());

        canvas.set_fill_style(&Colors::Edge.value().into());
        canvas.set_font("300 15px 'Exo 2'");

        canvas.set_text_align("left");
        let _result = canvas.fill_text("INSERT COIN", 220.0, 275.0);
        let _result = canvas.fill_text("PAUSE", 220.0, 325.0);
        let _result = canvas.fill_text("HUB", 220.0, 375.0);
        let _result = canvas.fill_text("SHOW EGDES", 220.0, 425.0);
        let _result = canvas.fill_text("STARFIELD", 220.0, 475.0);

        let _result = canvas.fill_text("ASTEROID", 600.0, 450.0);
        let _result = canvas.fill_text("UFO", 600.0, 475.0);
        let _result = canvas.fill_text("NEW LIFE", 600.0, 500.0);

        let _result = canvas.fill_text("10 POINTS", 745.0, 450.0);
        let _result = canvas.fill_text("50 POINTS", 745.0, 475.0);
        let _result = canvas.fill_text("3000 POINTS", 745.0, 500.0);

        canvas.set_text_align("right");
        let _result = canvas.fill_text("S", 475.0, 275.0);
        let _result = canvas.fill_text("P", 475.0, 325.0);
        let _result = canvas.fill_text("U", 475.0, 375.0);
        let _result = canvas.fill_text("E", 475.0, 425.0);
        let _result = canvas.fill_text("X", 475.0, 475.0);

        canvas.set_text_align("center");
        let _result = canvas.fill_text("BOOST", 710.0, 250.0);
        let _result = canvas.fill_text("TURN LEFT", 650.0, 320.0);
        let _result = canvas.fill_text("TURN RIGHT", 770.0, 320.0);
        let _result = canvas.fill_text("SHOOT", 710.0, 400.0);

        canvas.set_font("300 35px 'Exo 2'");
        let _result = canvas.fill_text("HELP", self.center_x, 210.0);

        canvas.restore();
    }
}
